File: js/test_assoc.js
const { ArgumentParser } = require('argparse');
const seedrandom = require('seedrandom');

const EqualityCheck = Object.freeze({
  ASSOCIATIVITY: "associativity",
  MULT_INVERSE: "mult_inverse",
  MULT_INVERSE_PI: "mult_inverse_pi",
});

function associativityTest(option) {
  const x = Math.random();
  const y = Math.random();
  const z = Math.random();
  if (option === EqualityCheck.ASSOCIATIVITY) {
    return x + (y + z) === (x + y) + z;
  } else if (option === EqualityCheck.MULT_INVERSE) {
    return (x * z) / (y * z) === x / y;
  } else if (option === EqualityCheck.MULT_INVERSE_PI) {
    const pi = Math.PI;
    return (x * z * pi) / (y * z * pi) === x / y;
  } else {
    throw new Error("Invalid option");
  }
}

function proportion(number, seedVal, option) {
  const rng = seedrandom(seedVal);
  let ok = 0;
  for (let i = 0; i < number; i++) {
    ok += associativityTest(option);
  }
  return (ok * 100) / number;
}

const parser = new ArgumentParser({
  description: 'Associativity test with seed.',
});

parser.add_argument('--seed', { type: Number, required: true, help: 'Seed value' });
parser.add_argument('--number', { type: Number, default: 10000, help: 'Number of tests' });
parser.add_argument('--equality-check', { choices: Object.values(EqualityCheck), default: EqualityCheck.ASSOCIATIVITY, help: 'Type of equality check (associativity, mult_inverse, mult_inverse_pi)' });
const args = parser.parse_args();

console.log(`${proportion(args.number, args.seed, args.equality_check)}%`);

File: swift/Sources/testassoc/testassoc.swift
import ArgumentParser
import Foundation
@main 
struct TestAssoc: ParsableCommand {
    //static let configuration = CommandConfiguration(abstract: "Equality test with seed.")
    
    @Option(name: .shortAndLong, help: "Seed value.")
    var seed: Int?
    
    @Option(name: .shortAndLong, help: "Number of tests.")
    var number: Int = 10000
    
    @Option(name: .shortAndLong, help: "Type of equality check.") // , completion: .values(EqualityCheck.allCases.map(\.rawValue)))
    var equalityCheck: EqualityCheck
    
    func run() throws {
        print("\(proportion(number: number, seedVal: seed ?? Int(Date().timeIntervalSince1970), equalityCheck: equalityCheck))%")
    }


    enum EqualityCheck: String, ExpressibleByArgument {
    case associativity = "associativity"
    case multInv = "mult-inverse"
    case multInvPi = "mult-inverse-pi"
}

func equalityTest(equalityCheck: EqualityCheck, x: Double, y: Double, z: Double) -> Bool {
    switch equalityCheck {
    case .associativity:
        return x + (y + z) == (x + y) + z
    case .multInv:
        return (x * z) / (y * z) == x / y
    case .multInvPi:
        return (x * z * Double.pi) / (y * z * Double.pi) == (x / y)
    }
}

func proportion(number: Int, seedVal: Int, equalityCheck: EqualityCheck) -> Int {
    var ok = 0
    srand(UInt32(seedVal))
    for _ in 0..<number {
        let x = Double(rand()) / Double(RAND_MAX)
        let y = Double(rand()) / Double(RAND_MAX)
        let z = Double(rand()) / Double(RAND_MAX)
        ok += equalityTest(equalityCheck: equalityCheck, x: x, y: y, z: z) ? 1 : 0
    }
    return ok * 100 / number
}
}









